Packages/vcs/Lib/vtk_ui/text.py

```diff
@@ -17,11 +19,118 @@ def __set_font(font, text_props):
         text_props.SetFontFamily(VTK_FONT_FILE)
         text_props.SetFontFile(font)
 
+
+def lum_normalize(component):
+    if component <= .03928:
+        return component / 12.92
+    else:
+        return ((component + .055) / 1.055) ** 2.4
+
+
+def luminance(color):
+    r, g, b = [lum_normalize(c) for c in color]
+    lum = .2126 * r + .7152 * g + .0722 * b
+    return lum
+
+
+def contrast_ratio(fg, bg):
+    lum_fg = luminance(fg)
+    lum_bg = luminance(bg)
+    l1 = max(lum_fg, lum_bg)
+    l2 = min(lum_fg, lum_bg)
+
+    return (l1 + .05) / (l2 + .05)
+
+
 def white_or_black(red, green, blue):
     """ Returns white or black to choose most contrasting color for provided color """
     # Convert to YIQ colorspace
-    yiq = (red * 255 * 299 + green * 255 * 587 + blue * 255 * 114) / 1000
-    return (0,0,0) if yiq >= 128 else (1, 1, 1)
+    lum = luminance((red, green, blue))
+    return (0, 0, 0) if lum >= .5 else (1, 1, 1)
+
+
+def contrasting_color(red, green, blue):
+    hue, saturation, value = rgb_to_hsv(red, green, blue)
+
+    #saturation /= 2.
+    r, g, b = red, green, blue
+    phi = .61803398875
+
+    iterations = 0
+    max_iters = 5
+
+    hsv = {"hue": hue, "value": value, "saturation": saturation}
+    var_keys = hsv.keys()
+    key = "value"
+
+    while contrast_ratio((red, green, blue), (r, g, b)) < 4.5:
```

Comment:
This is probably the only concerning part of this function; it may be theoretically possible to run through this loop forever, I'm not sure. I used some ideas from a couple of articles (mostly the Golden Ratio section [here](http://devmag.org.za/2012/07/29/how-to-choose-colours-procedurally-algorithms/)), and this seems to give pretty solid results. I could stick some kind of absolute maximum on how many iterations we're allowed and just give up and use the current color at that point... not sure.